internal/protoc/downloader.go

```diff
@@ -0,0 +1,329 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protoc
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/uber/prototool/internal/x/settings"
+	"github.com/uber/prototool/internal/x/vars"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type downloader struct {
+	logger    *zap.Logger
+	cachePath string
+	protocURL string
+	config    settings.Config
+
+	lock sync.RWMutex
+	// the looked-up and verified to exist base path
+	cachedBasePath string
+}
+
+func newDownloader(config settings.Config, options ...DownloaderOption) *downloader {
+	downloader := &downloader{
+		config: config,
+		logger: zap.NewNop(),
+	}
+	for _, option := range options {
+		option(downloader)
+	}
+	if downloader.config.Compile.ProtobufVersion == "" {
+		downloader.config.Compile.ProtobufVersion = vars.DefaultProtocVersion
+	}
+	return downloader
+}
+
+func (d *downloader) Download() (string, error) {
+	d.lock.RLock()
+	cachedBasePath := d.cachedBasePath
+	d.lock.RUnlock()
+	if cachedBasePath != "" {
+		return cachedBasePath, nil
+	}
+	return d.cache()
+}
+
+func (d *downloader) ProtocPath() (string, error) {
+	basePath, err := d.Download()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(basePath, "bin", "protoc"), nil
+}
+
+func (d *downloader) WellKnownTypesIncludePath() (string, error) {
+	basePath, err := d.Download()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(basePath, "include"), nil
+}
+
+func (d *downloader) Delete() error {
+	basePath, err := d.getBasePathNoVersion()
+	if err != nil {
+		return err
+	}
+	d.cachedBasePath = ""
+	d.logger.Debug("deleting", zap.String("path", basePath))
+	return os.RemoveAll(basePath)
+}
+
+func (d *downloader) cache() (string, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	basePath, err := d.getBasePath()
+	if err != nil {
+		return "", err
+	}
+	downloaded, err := d.isDownloaded(basePath)
+	if err != nil {
+		return "", err
+	}
+	if !downloaded {
+		if err := d.download(basePath); err != nil {
+			return "", err
+		}
+		if _, err := d.isDownloaded(basePath); err != nil {
+			return "", err
```

Comment:
I don't think this is reachable (see below).

Reply:
Updated.